src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp

```diff
@@ -431,7 +431,12 @@ struct ReplaceFileObjectStorageOperation final : public IDiskObjectStorageOperat
     {
         if (metadata_storage.existsFile(path_to))
         {
-            objects_to_remove = metadata_storage.getStorageObjects(path_to);
+            if (metadata_storage.getType() != MetadataStorageType::PlainRewritable
+                && metadata_storage.getType() != MetadataStorageType::Plain)
+            {
+                //  MetadataStorageFromPlainObjectStorageTransaction removes the source objects by itself.
+                objects_to_remove = metadata_storage.getStorageObjects(path_to);
```

Comment:
We have to consider hardlinks count before dropping the objects from file!

Reply:
Here you should drop remote blobs only when refcounter is 0, like it does `RemoveObjectStorageOperation`.
I made i quick fix just to prove it with tests. Details: https://github.com/ClickHouse/clickhouse-private/pull/39153#discussion_r2401447897